ow_plexil/src/plexil-adapter/FaultHierarchy.cpp

```diff
@@ -0,0 +1,309 @@
+        || subsystem_severity_high.empty() ){
+      std::cout << "WARNING: FAILED TO READ A SUBSYSTEM " << subsystem_name << "\n";
+      continue;
+    }
+
+    // Create and add new subsystem to m_subsystems
+    if (m_subsystems.find(subsystem_name) == m_subsystems.end()){
+      m_subsystems[subsystem_name] = Subsystem();
+      m_subsystems[subsystem_name].name = subsystem_name;
+      m_subsystems[subsystem_name].severity_threshold["Low"] = std::stoi(subsystem_severity_low);
+      m_subsystems[subsystem_name].severity_threshold["Medium"] = std::stoi(subsystem_severity_med);
+      m_subsystems[subsystem_name].severity_threshold["High"] = std::stoi(subsystem_severity_high);
+    }
+    else{
+      continue;
+    }
+ 
+    pugi::xml_node sys_affected_subsystems = subsystem.child("AffectedSubsystems");
+    pugi::xml_node faults = subsystem.child("Faults");
+
+    // get all affected subsystems and add them to their respective parent subsystem
+    for (pugi::xml_node affected_subsystem = sys_affected_subsystems.child("Sys"); affected_subsystem;
+         affected_subsystem = affected_subsystem.next_sibling("Sys")){
+      if (affected_subsystem.empty()){
+        std::cout << "WARNING: FAILED TO READ AN AFFECTED SUBSYSTEM " << affected_subsystem << "\n";
+        continue;
+      }
+      std::string affected_subsystem_name = affected_subsystem.attribute("Name").value();
+      std::string affected_subsystem_cascade_flag_str = affected_subsystem.attribute("Cascade").value();
+      bool cascade_flag = false;
+      // get cascade flag
+      if (affected_subsystem_cascade_flag_str == "True"){
+        cascade_flag = true;
+      }
+      else if (affected_subsystem_cascade_flag_str == "False"){
+        cascade_flag = false;
+      }
+      else{
+        std::cout << "WARNING: FAILED TO READ CASCADE FLAG ON AFFECTED SUBSYSTEM " << affected_subsystem << "\n";
+        continue;
+      }
+
+      m_subsystems[subsystem_name].affected_subsystems.push_back(std::make_pair(affected_subsystem_name, cascade_flag));
+    }
+
+    // Read all faults and add them to m_faults as well as their respective subsystems
+    for (pugi::xml_node fault = faults.child("Fault"); fault; fault = fault.next_sibling("Fault")){
+      std::string fault_name = fault.attribute("Name").value();
+      std::string fault_severity = fault.attribute("Severity").value();
+      if (fault_name.empty() || fault_severity.empty()){
+        std::cout << "WARNING: FAILED TO READ A FAULT IN SUBSYSTEM " << subsystem_name << "\n";
+        continue;
+      }
+
+      if (m_faults.find(fault_name) == m_faults.end()){
+        m_faults[fault_name] = Fault();
+        m_faults[fault_name].name = fault_name;
+      }
+      m_subsystems[subsystem_name].faults.push_back(fault_name);
+      m_faults[fault_name].affected_subsystems.push_back(std::make_pair(subsystem_name, fault_severity));
```

Comment:
I get this impression the if-statement is meant to ignore faults with duplicate names, but then the two lines after it go ahead and populate other vectors anyways. Due to those two lines I suspect having two faults of the same name will cause a bug because those lines are not included in the if-statement, but I am not sure.